# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [2]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv"

import pandas as pd

df = pd.read_csv(url)
df.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [3]:
df.dtypes

Unnamed: 0                         int64
Customer                          object
State                             object
Customer Lifetime Value          float64
Response                          object
Coverage                          object
Education                         object
Effective To Date                 object
EmploymentStatus                  object
Gender                            object
Income                             int64
Location Code                     object
Marital Status                    object
Monthly Premium Auto               int64
Months Since Last Claim          float64
Months Since Policy Inception      int64
Number of Open Complaints        float64
Number of Policies                 int64
Policy Type                       object
Policy                            object
Renew Offer Type                  object
Sales Channel                     object
Total Claim Amount               float64
Vehicle Class                     object
Vehicle Size    

In [5]:
df[(df["Total Claim Amount"] > 1000) & (df["Response"] == "Yes")].reset_index(drop=True)

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0          189  OK31456  California             11009.130490      Yes   
1          236  YJ16163      Oregon             11009.130490      Yes   
2          419  GW43195      Oregon             25807.063000      Yes   
3          442  IP94270     Arizona             13736.132500      Yes   
4          587  FJ28407  California              5619.689084      Yes   
..         ...      ...         ...                      ...      ...   
62       10351  FN44127      Oregon              3508.569533      Yes   
63       10373  XZ64172      Oregon             10963.957230      Yes   
64       10487  IX60941      Oregon              3508.569533      Yes   
65       10565  QO62792      Oregon              7840.165778      Yes   
66       10708  CK39096      Oregon              5619.689084      Yes   

    Coverage             Education Effective To Date EmploymentStatus Gender  \
0    Premium              Bachelor           1/24/11         Employed      F   
1    Premium              Bachelor           1/24/11         Employed      F   
2   Extended               College           2/13/11         Employed      F   
3    Premium                Master           2/13/11         Disabled      F   
4    Premium  High School or Below           1/26/11       Unemployed      M   
..       ...                   ...               ...              ...    ...   
62  Extended               College            1/5/11    Medical Leave      M   
63   Premium  High School or Below            2/8/11         Employed      M   
64  Extended               College            1/5/11    Medical Leave      M   
65  Extended               College           1/14/11         Employed      M   
66   Premium  High School or Below           1/26/11       Unemployed      M   

    ...  Number of Open Complaints Number of Policies     Policy Type  \
0   ...                        0.0                  1  Corporate Auto   
1   ...                        0.0                  1    Special Auto   
2   ...                        1.0                  2   Personal Auto   
3   ...                        0.0                  8   Personal Auto   
4   ...                        0.0                  1   Personal Auto   
..  ...                        ...                ...             ...   
62  ...                        1.0                  1   Personal Auto   
63  ...                        0.0                  1  Corporate Auto   
64  ...                        1.0                  1   Personal Auto   
65  ...                        2.0                  1   Personal Auto   
66  ...                        0.0                  1   Personal Auto   

          Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0   Corporate L3            Offer2          Agent         1358.400000   
1     Special L3            Offer2          Agent         1358.400000   
2    Personal L2            Offer1         Branch         1027.200000   
3    Personal L2            Offer1            Web         1261.319869   
4    Personal L1            Offer2            Web         1027.000029   
..           ...               ...            ...                 ...   
62   Personal L2            Offer2         Branch         1176.278800   
63  Corporate L2            Offer1          Agent         1324.800000   
64   Personal L3            Offer2         Branch         1176.278800   
65   Personal L3            Offer2          Agent         1008.000000   
66   Personal L3            Offer2            Web         1027.000029   

    Vehicle Class Vehicle Size Vehicle Type  
0      Luxury Car      Medsize          NaN  
1      Luxury Car      Medsize            A  
2      Luxury Car        Small            A  
3             SUV      Medsize            A  
4             SUV      Medsize            A  
..            ...          ...          ...  
62  Four-Door Car        Small          NaN  
63     Luxury SUV      Medsize          NaN 

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [7]:
# filter out the rows with response "No"
df_yes = df[df["Response"] == "Yes"].reset_index(drop=True)

#create a pivot with the average of total claim amount by policy type and gender   
pd.pivot_table(df_yes, values = "Total Claim Amount", index = "Policy Type", columns = "Gender", aggfunc = "mean").reset_index()


Gender     Policy Type           F           M
0       Corporate Auto  433.738499  408.582459
1        Personal Auto  452.965929  457.010178
2         Special Auto  453.280164  429.527942

In [ ]:
# the highest average claim amount is for "Personal" 
# females have higher averages than males in corporate and special policies types
# on personal policies claims have higher value for males


3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [26]:
# create a pivot with the total number of customers per state 
df_500 = pd.pivot_table(df, values = "Customer", index = "State", aggfunc = "count").reset_index()
df_500 

State  Customer
0     Arizona      1937
1  California      3552
2      Nevada       993
3      Oregon      2909
4  Washington       888

In [ ]:
#filter out states with less than 500 customers 
df_500 = df_500[df_500["Customer"] > 500]
df_500

State  Customer
0     Arizona      1937
1  California      3552
2      Nevada       993
3      Oregon      2909
4  Washington       888

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [38]:
# create a pivot with the max total customer lifetime value by education and gender 
pd.pivot_table(df, values = "Customer Lifetime Value", index = "Education", columns="Gender", aggfunc ="max").reset_index()


Gender             Education            F            M
0                   Bachelor  73225.95652  67907.27050
1                    College  61850.18803  61134.68307
2                     Doctor  44856.11397  32677.34284
3       High School or Below  55277.44589  83325.38119
4                     Master  51016.06704  50568.25912

In [ ]:
# create a pivot with the min total customer lifetime value by education and gender 
pd.pivot_table(df, values = "Customer Lifetime Value", index = "Education", columns="Gender", aggfunc ="min").reset_index()


Gender             Education            F            M
0                   Bachelor  1904.000852  1898.007675
1                    College  1898.683686  1918.119700
2                     Doctor  2395.570000  2267.604038
3       High School or Below  2144.921535  1940.981221
4                     Master  2417.777032  2272.307310

In [40]:
# create a pivot with the median total customer lifetime value by education and gender 
pd.pivot_table(df, values = "Customer Lifetime Value", index = "Education", columns="Gender", aggfunc = "median").reset_index()


Gender             Education            F            M
0                   Bachelor  5640.505303  5548.031892
1                    College  5623.611187  6005.847375
2                     Doctor  5332.462694  5577.669457
3       High School or Below  6039.553187  6286.731006
4                     Master  5729.855012  5579.099207

max value for total lifetime value for females has a bachelor degree, while for males has high school or below education level
min value for total lifetime value for females has college degree. For males has a bachelor degree with a value of 1898
the median value for a female with high school or below education level is the highest in females.
there are half of males with a college degree that have a total lifetime value higher than 6005,85

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here